MalformedInputException on Locale C.UTF-8 on JDK 11

Coming from https://github.com/oshi/oshi/pull/2525#issuecomment-1838992663

1. My library prints output to stdout containing the [degree sign, UTF-8 0xC2B0](https://www.fileformat.info/info/unicode/char/b0/index.htm).
2. This output is printed as part of our CI testing.  One such CI is on a Solaris 11.4 x86 VM.  Important to this discussion, the locale is set to `LANG=C.UTF-8`.
3. While the CI test passes, the output catches the same error we saw in #95 on Windows

Full log: https://github.com/oshi/oshi/actions/runs/7089320523/job/19293798675
Excerpt:
```
  Caused by: java.lang.AssertionError: Unexpected exception caught!
  	at de.sormuras.junit.platform.maven.plugin.JUnitPlatformMojo.execute(JUnitPlatformMojo.java:433)
<snip>
  Caused by: java.io.UncheckedIOException: java.nio.charset.MalformedInputException: Input length = 1
  	at java.base/java.nio.file.FileChannelLinesSpliterator.readLine(FileChannelLinesSpliterator.java:176)
  	at java.base/java.nio.file.FileChannelLinesSpliterator.forEachRemaining(FileChannelLinesSpliterator.java:116)
  	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
  	at de.sormuras.junit.platform.maven.plugin.JavaExecutor.evaluate(JavaExecutor.java:147)
<snip>
  Caused by: java.nio.charset.MalformedInputException: Input length = 1
  	at java.base/java.nio.charset.CoderResult.throwException(CoderResult.java:274)
  	at java.base/sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:339)
  	at java.base/sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
  	at java.base/java.io.BufferedReader.fill(BufferedReader.java:161)
  	at java.base/java.io.BufferedReader.readLine(BufferedReader.java:326)
  	at java.base/java.io.BufferedReader.readLine(BufferedReader.java:392)
  	at java.base/java.nio.file.FileChannelLinesSpliterator.readLine(FileChannelLinesSpliterator.java:174)
```
The code involved (exception thrown in `Files.lines()`
```java
if (captureIO) {
  String encoding = System.getProperty("native.encoding"); // Populated on Java 18 and later
  Charset charset = encoding != null ? Charset.forName(encoding) : Charset.defaultCharset();
  try (Stream<String> stdoutput = Files.lines(outputPath, charset);
      Stream<String> erroutput = Files.lines(errorPath, charset)) {
```
Since this is pre-JDK 18, the fix for #95 using the "native.encoding" doesn't apply, however, the symptom is the same. The charset being used doesn't understand one of the characters being processed.

This [Stack Overflow answer](https://stackoverflow.com/a/26268235/1161484) suggests trying multiple charsets.  [Another answer](https://stackoverflow.com/a/44233948/1161484) claims 
> ISO-8859-1 is an all-inclusive charset, in the sense that it's guaranteed not to throw MalformedInputException. So it's good for debugging

Not sure if this is something you want to handle here, but at a minimum I'd request that you at least catch the exception and provide a better error message indicating that the default charset (print it out) can't handle a line of output.

--- a/README.md
+++ b/README.md
@@ -235,6 +235,16 @@ Duration between output and error log file sizes during execution (JAVA executio
 </configuration>
 ```
 
+### Log Charset
+
+Charset format for the output and error log files. Defaults to `Charset.defaultCharset()` for JDK 17 and lower, `System.getProperty("native.encoding")` for JDK 18 and higher.
+
+```xml
+<configuration>
+  <charset>UTF-8</charset>
+</configuration>
+```
+
 ### Isolation Level
 
 `ClassLoader` hierarchy configuration.

--- a/pom.xml
+++ b/pom.xml
@@ -388,6 +388,7 @@
                                 <!--<pomInclude>multiple-engines-*</pomInclude>-->
                                 <!--<pomInclude>timeout</pomInclude>-->
                                 <!--<pomInclude>executionProgress</pomInclude>-->
+                                <!--<pomInclude>charset</pomInclude>-->
                             </pomIncludes>
                             <pomExcludes>
                                 <!-- None, B. -->

--- a/src/it/charset/pom.xml
+++ b/src/it/charset/pom.xml
@@ -1,0 +1,36 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>it</groupId>
+    <artifactId>setup</artifactId>
+    <version>0</version>
+  </parent>
+
+  <artifactId>execution-progress</artifactId>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-api</artifactId>
+      <version>${junit.jupiter.version}</version>
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>@project.groupId@</groupId>
+        <artifactId>@project.artifactId@</artifactId>
+        <configuration>
+          <executor>JAVA</executor>
+          <charset>ISO-8859-1</charset>
+        </configuration>
+      </plugin>
+    </plugins>
+  </build>
+
+</project>

--- a/src/it/charset/src/test/java/CharsetTests.java
+++ b/src/it/charset/src/test/java/CharsetTests.java
@@ -1,0 +1,9 @@
+import org.junit.jupiter.api.Test;
+
+class CharsetTests {
+
+  @Test
+  void outputSomething() throws Exception {
+    System.out.println("UTF-8 degree sign is °");
+  }
+}

--- a/src/it/charset/src/test/java/CharsetTests.java
+++ b/src/it/charset/src/test/java/CharsetTests.java
@@ -4,6 +4,8 @@ class CharsetTests {
 
   @Test
   void outputSomething() throws Exception {
+    // The degree sign (hex b0) will be prepended by 2c in UTF-8
+    // For ISO output check for optional Â (hex 2c) character
     System.out.println("UTF-8 degree sign is °");
   }
 }

--- a/src/it/charset/verify.bsh
+++ b/src/it/charset/verify.bsh
@@ -19,7 +19,7 @@ verifier.verifyLogMatches(new String[] {
   ">> BEGIN >>",
   "[INFO] Launching JUnit Platform " + junitPlatformVersion + "...",
   ">> Platform executes tests...>>",
-  "\\Q[INFO]\\E UTF-8 degree sign is Â°",
+  "\\Q[INFO]\\E UTF-8 degree sign is Â?°",
   ">> Platform executes tests...>>",
   "[INFO] BUILD SUCCESS",
   ">> END. >>"

--- a/src/it/charset/verify.bsh
+++ b/src/it/charset/verify.bsh
@@ -1,0 +1,28 @@
+import it.Verifier;
+
+Verifier verifier = new Verifier(basedir.toPath());
+
+// verifier.verifyBadLines(); // warning and error lines are expected
+
+verifier.verifyReadable(new String[] {
+  "pom.xml",
+  "src/test/java/CharsetTests.java",
+  "target/test-classes/CharsetTests.class"
+});
+
+verifier.verifyNotExists(new String[] {
+  "target/classes",
+  "target/surefire-reports"
+});
+
+verifier.verifyLogMatches(new String[] {
+  ">> BEGIN >>",
+  "[INFO] Launching JUnit Platform " + junitPlatformVersion + "...",
+  ">> Platform executes tests...>>",
+  "\\Q[INFO]\\E UTF-8 degree sign is Â°",
+  ">> Platform executes tests...>>",
+  "[INFO] BUILD SUCCESS",
+  ">> END. >>"
+});
+
+return verifier.isOk();

--- a/src/main/java/de/sormuras/junit/platform/maven/plugin/JUnitPlatformMojo.java
+++ b/src/main/java/de/sormuras/junit/platform/maven/plugin/JUnitPlatformMojo.java
@@ -92,6 +92,10 @@ public class JUnitPlatformMojo extends AbstractMavenLifecycleParticipant impleme
   @Parameter(defaultValue = "300")
   private long timeout = 300L;
 
+  /** Charset for log files. */
+  @Parameter(defaultValue = "")
+  private String charset = null;
+
   /** Execution progress update duration in seconds. */
   @Parameter(defaultValue = "60")
   private long executionProgress = 60;
@@ -506,6 +510,10 @@ public class JUnitPlatformMojo extends AbstractMavenLifecycleParticipant impleme
 
   long getTimeout() {
     return timeout;
+  }
+
+  String getCharset() {
+    return charset;
   }
 
   long getExecutionProgress() {

--- a/src/main/java/de/sormuras/junit/platform/maven/plugin/JavaExecutor.java
+++ b/src/main/java/de/sormuras/junit/platform/maven/plugin/JavaExecutor.java
@@ -19,7 +19,6 @@ import de.sormuras.junit.platform.isolator.Modules;
 import de.sormuras.junit.platform.isolator.TestMode;
 import java.io.File;
 import java.io.IOException;
-import java.nio.charset.CharacterCodingException;
 import java.nio.charset.Charset;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -102,7 +101,6 @@ class JavaExecutor {
     mojo.debug("");
     mojo.debug("Starting process...");
     cmd.forEach(mojo::debug);
-    int exitValue = -1;
     try {
       Process process = builder.start();
       // Java 11 debug("Process started: #%d %s", process.pid(), process.info());
@@ -139,29 +137,28 @@ class JavaExecutor {
         }
         return -2;
       }
-      exitValue = process.exitValue();
+      int exitValue = process.exitValue();
       if (captureIO) {
         String encoding = mojo.getCharset();
         if (encoding == null) {
           encoding = System.getProperty("native.encoding"); // Populated on Java 18 and later
         }
         Charset charset = encoding != null ? Charset.forName(encoding) : Charset.defaultCharset();
-        try (Stream<String> stdoutput = Files.lines(outputPath, charset);
-            Stream<String> erroutput = Files.lines(errorPath, charset)) {
-          stdoutput.forEach(exitValue == 0 ? mojo::info : mojo::error);
-          erroutput.forEach(exitValue == 0 ? mojo::warn : mojo::error);
+        try {
+          try (Stream<String> stdoutput = Files.lines(outputPath, charset);
+              Stream<String> erroutput = Files.lines(errorPath, charset)) {
+            stdoutput.forEach(exitValue == 0 ? mojo::info : mojo::error);
+            erroutput.forEach(exitValue == 0 ? mojo::warn : mojo::error);
+          }
         } catch (IOException e) {
           mojo.warn("Reading output/error logs failed: {0}", e);
         }
       }
-    } catch (CharacterCodingException cce) {
-      // Possibly thrown from Files.lines and not caught above
-      mojo.warn("Charset error reading output/error logs: {0}", cce);
+      return exitValue;
     } catch (IOException | InterruptedException e) {
       mojo.error("Executing process failed: {0}", e);
       return -1;
     }
-    return exitValue;
   }
 
   // Supply standard options for Java foundation tool

--- a/src/main/java/de/sormuras/junit/platform/maven/plugin/JavaExecutor.java
+++ b/src/main/java/de/sormuras/junit/platform/maven/plugin/JavaExecutor.java
@@ -19,6 +19,7 @@ import de.sormuras.junit.platform.isolator.Modules;
 import de.sormuras.junit.platform.isolator.TestMode;
 import java.io.File;
 import java.io.IOException;
+import java.nio.charset.CharacterCodingException;
 import java.nio.charset.Charset;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -152,11 +153,14 @@ class JavaExecutor {
           mojo.warn("Reading output/error logs failed: {0}", e);
         }
       }
-      return exitValue;
+    } catch (CharacterCodingException cce) {
+      // Possibly thrown from Files.lines and not caught above
+      mojo.warn("Charset error reading output/error logs: {0}", cce);
     } catch (IOException | InterruptedException e) {
       mojo.error("Executing process failed: {0}", e);
       return -1;
     }
+    return exitValue;
   }
 
   // Supply standard options for Java foundation tool

--- a/src/main/java/de/sormuras/junit/platform/maven/plugin/JavaExecutor.java
+++ b/src/main/java/de/sormuras/junit/platform/maven/plugin/JavaExecutor.java
@@ -139,7 +139,10 @@ class JavaExecutor {
       }
       int exitValue = process.exitValue();
       if (captureIO) {
-        String encoding = System.getProperty("native.encoding"); // Populated on Java 18 and later
+        String encoding = mojo.getCharset();
+        if (encoding == null) {
+          encoding = System.getProperty("native.encoding"); // Populated on Java 18 and later
+        }
         Charset charset = encoding != null ? Charset.forName(encoding) : Charset.defaultCharset();
         try (Stream<String> stdoutput = Files.lines(outputPath, charset);
             Stream<String> erroutput = Files.lines(errorPath, charset)) {

--- a/src/main/java/de/sormuras/junit/platform/maven/plugin/JavaExecutor.java
+++ b/src/main/java/de/sormuras/junit/platform/maven/plugin/JavaExecutor.java
@@ -102,6 +102,7 @@ class JavaExecutor {
     mojo.debug("");
     mojo.debug("Starting process...");
     cmd.forEach(mojo::debug);
+    int exitValue = -1;
     try {
       Process process = builder.start();
       // Java 11 debug("Process started: #%d %s", process.pid(), process.info());
@@ -138,7 +139,7 @@ class JavaExecutor {
         }
         return -2;
       }
-      int exitValue = process.exitValue();
+      exitValue = process.exitValue();
       if (captureIO) {
         String encoding = mojo.getCharset();
         if (encoding == null) {

--- a/src/test/java/de/sormuras/junit/platform/maven/plugin/JUnitPlatformMojoTests.java
+++ b/src/test/java/de/sormuras/junit/platform/maven/plugin/JUnitPlatformMojoTests.java
@@ -34,6 +34,7 @@ class JUnitPlatformMojoTests {
     assertEquals(Isolation.NONE, mojo.getIsolation());
     assertEquals(300L, mojo.getTimeout());
     assertEquals(60L, mojo.getExecutionProgress());
+    assertNull(mojo.getCharset());
 
     assertNotNull(mojo.getLog());
     assertNull(mojo.getMavenProject());

